fairscale/nn/misc/checkpoint_activations.py

```diff
@@ -63,25 +64,55 @@ def checkpoint_wrapper(module: nn.Module, offload_to_cpu: bool = False) -> nn.Mo
         (nn.Module):
             wrapped module
     """
+    # Build a list of BN classes and use two functions only make them track
+    # running stats in one of the forward function.
+    bn_list = []
+    for name, child in module.named_modules():
+        # _BatchNorm is base for bn1d, bn2d, bn3d and sync_bn, apex_sync_bn
+        if isinstance(child, _BatchNorm):
+            bn_list.append([child, None])
+
+    def pre_forward() -> None:
+        for bn_and_bool in bn_list:
+            bn_and_bool[1] = bn_and_bool[0].track_running_stats  # type: ignore
+            bn_and_bool[0].track_running_stats = False  # type: ignore
+
+    def post_forward() -> None:
+        for bn_and_bool in bn_list:
+            assert bn_and_bool[1] is not None
+            bn_and_bool[0].track_running_stats = bn_and_bool[1]  # type: ignore
```

Comment:
This actually seems like a general problem, I wonder if we can make something more standalone/reusable? In particular, I feel that BatchNorm layers should not update stats when in `torch.no_grad()` mode. Currently they only respect `model.eval()`, but perhaps we can have a simple module-level wrapper that registers the relevant hooks?

Something like (untested):
```python
def patch_batch_norm(module: nn.Module):
    if isinstance(module, _BatchNorm):
        module.register_forward_pre_hook(bn_pre_hook)
        module.register_forward_hook(bn_post_hook)
    return module

def bn_pre_hook(module, input):
    if not torch.is_grad_enabled():
        module._backup_track_running_stats = module.track_running_stats
        module.track_running_stats = False

def bn_post_hook(module, input, output):
    if not torch.is_grad_enabled():
        module.track_running_stats = module._backup_track_running_stats
```

What do you think?

Reply:
Great idea! Thanks for suggesting. Let me give it a try. I am a bit worries that users may want to keep the behavior of tracking stats under no_grad() but that's probably a very rare thing if at all possible. The current method only affect checkpointed fwd pass.